src/toxcore/messenger/packet/file_send_request.rs

```diff
@@ -0,0 +1,140 @@
+/*! FileSendRequest struct.
+It is used to start transferring file to a friend.
+*/
+
+use nom::{rest, le_u8, be_u32, be_u64};
+
+use std::str;
+
+use crate::toxcore::binary_io::*;
+use crate::toxcore::crypto_core::*;
+
+/// Type of file to transfer
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum FileType {
+    /// Normal data file.
+    Data = 0,
+    /// Avatar image file.
+    Avatar,
+}
+
+/// Maximum file name size in bytes
+const MAX_FILESEND_FILENAME_LENGTH: usize = 255;
+
+impl FromBytes for FileType {
+    named!(from_bytes<FileType>,
+        switch!(be_u32,
+            0 => value!(FileType::Data) |
+            1 => value!(FileType::Avatar)
+        )
+    );
+}
+
+/** FileSendRequest is a struct that holds info to start transferring file to a friend.
+
+This packet is used to start transferring sender's file to a friend.
+`file_type` and `file_size` are sent in big endian format.
+
+Serialized form:
+
+Length    | Content
+--------- | ------
+`1`       | `0x50`
+`1`       | `file_id`
+`4`       | `file_type`(0 = normal file, 1 = avatar file)
+`8`       | `file_size`
+`32`      | `file_unique_id`(a PrecomputedKey)
```

Comment:
How do you calculate file_unique_id? Why is it a PrecomputedKey?

Reply:
`file_unique_id` is for restarting of client process while transferring files.
When client restarts, it can find the file which was transferring using this `file_unique_id`.
It is a random symmetric key.
https://github.com/TokTok/c-toxcore/blob/0aad180d1e1d7dae7cbf2868e139af0a4dc34fe7/toxcore/tox.c#L1395

Reply:
```
    uint8_t f_id[FILE_ID_LENGTH];

    if (!file_id) {
        /* Tox keys are 32 bytes like FILE_ID_LENGTH. */
        new_symmetric_key(f_id);
        file_id = f_id;
    }
/* Fill a key CRYPTO_SYMMETRIC_KEY_SIZE big with random bytes */
void new_symmetric_key(uint8_t *key)
{
    random_bytes(key, CRYPTO_SYMMETRIC_KEY_SIZE);
}
```

So it's not a PrecomputedKey, it is a uniq 32-byte random key.

Reply:
> It is a random symmetric key.
It is not a random symmetric key. It is a random key of 32 bytes

Reply:
done